libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/ParametricSpec.hs

```diff
@@ -0,0 +1,301 @@
+import Cardano.Ledger.Credential (Credential, StakeReference (..))
+import Cardano.Ledger.Crypto (Crypto)
+import Cardano.Ledger.Keys (KeyHash, KeyRole (..))
+import Cardano.Ledger.Mary (Mary, MaryValue)
+import Cardano.Ledger.PoolParams (PoolParams (ppId))
+import Cardano.Ledger.Shelley (Shelley)
+import Cardano.Ledger.Shelley.LedgerState (AccountState (..), StashedAVVMAddresses)
+import Cardano.Ledger.Shelley.TxCert
+import Cardano.Ledger.Shelley.TxOut (ShelleyTxOut (..))
+import Constrained hiding (Value)
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Word (Word64)
+import Test.Cardano.Ledger.Constrained.Conway.Cert (conwayTxCertSpec, shelleyTxCertSpec)
+import Test.Cardano.Ledger.Constrained.Conway.Deleg (EraDeleg (..))
+import Test.Cardano.Ledger.Constrained.Conway.Instances (ConwayFn, IsConwayUniv, toDelta_)
+import Test.Cardano.Ledger.Constrained.Conway.SimplePParams
+import qualified Test.Cardano.Ledger.Constrained.Conway.SimplePParams as SimplePParams
+
+-- ===========================================================
+
+-- | The class EraOut supports Era parametric Specifications that
+--   primarily navigate the differences in types parameterized type Family TxOut.
+--   Additional support for phased out Type Families like InstantaneousRewards,
+--   GenDelegs, FutureGenDelegs, StashedAVVMAddresses, and Ptrs, are also provided
+class
+  ( HasSpec fn (TxOut era)
+  , IsNormalType (TxOut era)
+  , HasSpec fn (StashedAVVMAddresses era)
+  , EraTxOut era
+  , IsConwayUniv fn
+  , EraPP era
+  , EraDeleg era
+  ) =>
+  EraOut era fn
+  where
+  irewardSpec :: Term fn AccountState -> Specification fn (InstantaneousRewards (EraCrypto era))
+  hasPtrs :: proxy era -> Term fn Bool
+  correctTxOut ::
+    Term fn (Map (Credential 'Staking (EraCrypto era)) (KeyHash 'StakePool (EraCrypto era))) ->
+    Term fn (TxOut era) ->
+    Pred fn
+
+betterTxOutShelley ::
+  (EraTxOut era, Value era ~ Coin, IsConwayUniv fn) =>
+  Term fn (Map (Credential 'Staking (EraCrypto era)) (KeyHash 'StakePool (EraCrypto era))) ->
+  Term fn (ShelleyTxOut era) ->
+  Pred fn
+betterTxOutShelley delegs txOut =
+  match txOut $ \ [var|addr|] [var|val|] ->
+    [ match val $ \ [var|c|] -> [0 <. c, c <=. fromIntegral (maxBound :: Word64)]
+    , (caseOn addr)
+        ( branch $ \ [var|network|] _ [var|stakeref|] ->
+            [ assert $ network ==. lit Testnet
+            , satisfies stakeref (delegatedStakeReference delegs)
+            ]
+        )
+        ( branch $ \bootstrapAddr ->
+            match bootstrapAddr $ \_ [var|nm|] _ ->
+              (caseOn nm)
+                (branch $ \_ -> False)
+                (branch $ \_ -> True)
+        )
+    ]
+
+betterTxOutMary ::
+  (EraTxOut era, Value era ~ MaryValue (EraCrypto era), IsConwayUniv fn) =>
+  Term fn (Map (Credential 'Staking (EraCrypto era)) (KeyHash 'StakePool (EraCrypto era))) ->
+  Term fn (ShelleyTxOut era) ->
+  Pred fn
+betterTxOutMary delegs txOut =
```

Comment:
These specs seem to have a lot of constraints just copy-pasted over, I think these constraints should be factored out into a separate function to be reused across the specs

Reply:
This is difficult, because two things are varying, Value and TxOut, so it is hard to make a general type that captures all the variations. I will think further on this

Reply:
I am going to leave this for another PR